the-state-of-doctests-in-haskell/article.md

```diff
@@ -0,0 +1,253 @@
+[`doctest`](https://en.wikipedia.org/wiki/Doctest) module, which is included in the [`Python programming language`](https://en.wikipedia.org/wiki/Python_(programming_language)).
+
+Documenting software is extremely challenging.
+The doctests concept makes the documentation process not only pleasant and effective - it actually allows Haskell functions to speak for themselves - but
+also helps to get the most out of your testing efforts.
+
+For the sake of simplicity and convenience, we won't cover the markup used to define a doctest.
+Instead, we'll try to go over various libraries the community uses to run the doctest examples.
+Particularly, we'll describe a brand new [`cabal-docspec`](https://github.com/phadej/cabal-extras/blob/master/cabal-docspec/MANUAL.md)
+library in a little bit more details.
+
+### Creating a demo project containing doctests
+
+Let's create a project with short and simple haddock documentation containing doctests:
+
+1. Bootstrap a project using [`stack`](https://docs.haskellstack.org/en/stable/README/) first:
+
+      `stack new doctests-demo`
+
+2. Go to the project's root directory and create a [Haskell module](https://www.haskell.org/tutorial/modules.html):
+
+      `cd doctests-demo`
+
+      `touch src/Sample.hs`
+
+3. Add some dummy functions with doctests to newly created `Sample.hs` module:
+
+      ```haskell
+
+        -- src/Sample.hs
+
+        module Sample where
+
+        -- |
+        -- >>> foo + 13
+        -- 55
+        foo :: Integer
+        foo = 42
+
+        -- |
+        -- >>> bar
+        -- "bar"
+        bar :: String
+        bar = "bar"
+
+      ```
+
+4. Next, we'll be launching those doctest examples using various libraries in the following sections.
+
+### Libraries walkthrough
+
+#### `doctest`
+
+[doctest](https://github.com/sol/doctest) is probably one of the most commonly used and actively maintained libraries.
+
+Follow the steps below in order to launch the doctest examples in our `doctests-demo` project:
+
+1. Install the library via `stack`:
+
+    `stack install doctest`
+
+2. Go to the project's root directory and use the library's executable:
+
+    `cd doctests-demo`
+
+    `doctest src`
+
+3. The command above should output something like the following:
+
+    `Examples: 2  Tried: 2  Errors: 0  Failures: 0`
+
+While being an awesome software, the library has some limitations and drawbacks.
+Firstly, it has a dependency on [ghc as a library](https://wiki.haskell.org/GHC/As_a_library) - it's very likely that something eventually breaks down when
+you try to switch the project to a new compiler version.
+Secondly, when used via [`stack`](https://docs.haskellstack.org/en/stable/README/), one might need to create an additional test suite to bring the project
+dependencies into the scope:
+
+* Let's edit our doctest examples a little (don't forget to add `aeson` and `text` packages to the list of project dependencies):
```

Comment:
I would use an HTML spoiler tag with the name "example"

Reply:
https://www.w3schools.com/tags/tag_details.asp

Reply:
I'm not sure I agree with these changes, the examples are short and I think they're best left auto-expanded.

Reply:
Yeah, facts. No idea why I commented that.